Deploy astropy units throughout radis : Take I

ritzvik opened the pull request:
Initial steps towards #54 @erwanp 

See if the current implementation is fine or should I proceed in some another way ?
I was trying to fix `SpectrumFactory` first as it is easy to understand and therefore made `get_waverange` units-aware. 

Kindly see the changes in `lbl/test_base.py`. Wavelengths are provided in `cm` and converted to `nm` internally.

If the solution is acceptable, I would proceed with fixing all parameters of `SpectrumFactory`.

ritzvik commented:
I will also change the docstrings when this is given green light :)

erwanp commented:
Hi @ritzvik !

Just saw the example in `test_base.py`: well done !

I have two comments about it:

1. I think we shouldn't mix the different tests. 
What about making a new one in `test_factory.py`, for instance based on the sample below:

```
@pytest.mark.fast
def test_waverange_units_conversion(verbose=True, warnings=True, *args, **kwargs):
    '''... 

    '''

    if verbose: printm('>>> test_waverange_units_conversion')

    setup_test_line_databases()  # add HITRAN-CO-TEST in ~/.radis if not there
    # >> This test has CO lines in the 4300 - 4666 nm range 


    # Different Tests:
    input_wavelengths = [(4300, 4500),
                         (4300 * u.nm, 4500 * u.nm),
                         (4.3 * u.µm, 4.5 * u.µm),  
                         ... other tests] 
    expected_wavelengths = [(4300, 4500),
                            (4300, 4500), 
                            (4300, 4500),
                          ... other tests]

    
    # Run the tests
    for (wlmin, wlmax), (expected_wlmin, expected_wlmax) in zip(
            input_wavelengths, expected_wavelengths):
    
        sf = SpectrumFactory(wavelength_min=4300, wavelength_max=4666,
                             wstep=0.01,
                             cutoff=1e-30,
                             path_length=1, 
                             mole_fraction=1,
                             isotope=[1], 
                             verbose=verbose)
        s = sf.eq_spectrum(Tgas=300)
        
        assert s.get_wavelength()[0] == expected_wlmin
        assert s.get_wavelength()[-1] == expected_wlmax

    return
```

2. We should still have a way to deal with inverse centimeters (cm^-1), which is the standard unit in infrared measurements, but is not a default `astropy` unit. Do you know how astropy users do?

``` 
1000 * 1/u.cm ? 
(1/1000) * u.cm ?   # like in your test_base example
```

Typically, we'll want people to run calculations in cm^-1 or m^-1, and convert between these two units

erwanp commented:
@ritzvik  we can also discuss on [Gitter](https://gitter.im/radis-radiation/community) if more convenient!

ritzvik commented on radis/phys/units_astropy.py:
```diff
@@ -0,0 +1,32 @@
+import astropy.units as u
+
+
+def strip_units(quantity, units=None):
```
I’ll do the these changes.
I am away from my system right now.

ritzvik replied:
One more thing.
`convert_and_strip_units` would be a better name in my opinion, as units are first converted and then stripped.

erwanp commented:
I also see your PR as an opportunity to get rid of the 4 inputs: `wavelength_min, wavelength_max, wavenumber_min, wavenumber_max` and replace them with only 2 : `wmin, wmax` (for instance), with one third `wunit='cm-1'` argument when people don't use Astropy units in `wmin, wmax` directly. 

We'd keep the old parameters available for some time, with a DeprecationWarning, then delete them at some point.

ritzvik commented:
> I also see your PR as an opportunity to get rid of the 4 inputs: `wavelength_min, wavelength_max, wavenumber_min, wavenumber_max` and replace them with only 2 : `wmin, wmax` (for instance), with one third `wunit='cm-1'` argument when people don't use Astropy units in `wmin, wmax` directly.
> 
> We'd keep the old parameters available for some time, with a DeprecationWarning, then delete them at some point.

This I would do this in a separate PR. In this PR I would like to concentrate on changing the parameters of `SpectrumFactory`. Is that okay with you?
Maybe open an issue regarding this.

erwanp commented:
Perfect let's do!

ritzvik commented:
@erwanp I did the suggested changes in the latest commit a97a8c4 .
Regarding tests, I left the previous one unchanged and made a copy of the same test with different values for `wavelength_min` and `wavelength_max`

erwanp commented:
Great! I didn't know the `@pytest.mark.parametrize` trick. 

However, the test does not actually test if the wavelengths units are correctly read. It's still testing the convergence of an optically thick towards the Planck function. 
Hence the 
```
        
        assert np.isclose(s.get_wavelength()[0], expected_wlmin)
        assert np.isclose(s.get_wavelength()[-1], expected_wlmax)
```
in my suggestion above. 

Also, using HITRAN-CO rather than HITEMP-CO2 just makes the best much faster because there are much fewer lines. Could you run it on your system?

ritzvik commented:
Now 

> Great! I didn't know the `@pytest.mark.parametrize` trick.
> 
> However, the test does not actually test if the wavelengths units are correctly read. It's still testing the convergence of an optically thick towards the Planck function.
> Hence the
> 
> ```
>         
>         assert np.isclose(s.get_wavelength()[0], expected_wlmin)
>         assert np.isclose(s.get_wavelength()[-1], expected_wlmax)
> ```
> 
> in my suggestion above.
> 
> Also, using HITRAN-CO rather than HITEMP-CO2 just makes the best much faster because there are much fewer lines. Could you run it on your system?

Ok I would need some time to comprehend this, but I will do it.

ritzvik commented:
So, as of now, the status of this PR

Things Done:
  - `wavenum_min`, `wavenum_max`, `wavelength_min`, `wavelength_max`, `pressure` and `path_length` parameters in `SpectrumFactory` made unit-aware. 
  - `get_waverange()` in `lbl/base.py` made unit-aware.
  - Added astropy as a dependency
  - Updated the docstrings
  - Added astropy intersphinx
  - Replicated an already present test with astropy-unit quantities as parameters

Things I couldn't do:
  - `Tref` in `SpectrumFactory` can't be made unit-aware because conversion between Celsius, Fahrenheit and Kelvin is done differently than other quantities. Some problem in `astropy.units`. The conversion function is not documented properly.
  - I couldn't get the test running you mentioned above. Therefore whatever I have written, I have commented out. Please edit the tests according to your convenience. I am sorry about this :(

Lastly, should I squash or rearrange the commits?

ritzvik commented:
Hi @erwanp 

I've also made `Tref` units-aware although, it needed some workaround. I've also rearranged the commits.

Coming to the matter of tests, I believe we can now drop `test_optically_thick_limit_1iso`, because `test_optically_thick_limit_1iso_different_params` is doing the same thing along with checking unit-convertibility . 
Also, I couldn't get the test running which you mentioned in https://github.com/radis/radis/pull/55#issuecomment-573308075 , therefore, I commented out whatever I wrote. Kindly see if you can fix that.

Other than that, I guess this PR is pretty much complete.

erwanp commented:
Hello! I found an error when using wavenumbers as inputs: the `1/u.cm` is not a UnitBase else it would crash. I modified your code to also allow for Quantity.

In the process I also:
- added base tests for the `convert_and_strip_units` function
- replaced the SpectrumFactory test with the fast tests from https://github.com/radis/radis/pull/55#issuecomment-573308075
- added a default rounding at 10 digits because 4.5 µm would become 4499.99999999999 
- found about the `equivalencies=u.spectral()` option, which is amazing because we can now give wavenumbers in cm-1, Hz, or even nm. First step towards #56 !

What do you think?

ritzvik commented on radis/phys/units_astropy.py:
```diff
@@ -0,0 +1,59 @@
+    if isinstance(quantity, u.Quantity):
+        if output_unit in (u.deg_C, u.imperial.deg_F, u.K):
+            quantity = quantity.to_value(output_unit, equivalencies = u.temperature())
+        elif isinstance(output_unit, u.UnitBase) or isinstance(output_unit, u.Quantity):
```
Nothing significant, but can we change this `elif` statement to `elif isinstance(output_unit, [u.UnitBase, u.Quantity]):`. 
I know it seems trivial but it would contribute to code cleanliness in the long run.

erwanp commented:
Nice! I added some extra final touches:

- added test for temperature unit conversions  [ in SpectrumFactory] 
- added unit conversion in eq_spectrum, noneq_spectrum methods of SpectrumFactory and BandFactory   (that was important! these functinos are called after SpectrumFactory initialization for batch processing cases)

And the docs: 
- mentioned units in docstrings of calc_spectrum   [ works well already, it just forwards arguments to SpectrumFactory] 
- mentioned units in online docs and 2nd main example
